notebooks/notebook_template_review.py

```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
+def parse_dir(directory: str) -> int:
     """
         Recursively walk the specified directory, reviewing each notebook (.ipynb) encountered.
         
-        directory: The directory path.
+            directory: The directory path.
+            
+        Returns the numbern of errors
     """
+    exit_code = 0
+    
     entries = os.scandir(directory)
     for entry in entries:
         if entry.is_dir():
             if entry.name[0] == '.':
                 continue
             if entry.name == 'src' or entry.name == 'images' or entry.name == 'sample_data':
                 continue
-            print("\n##", entry.name, "\n")
-            parse_dir(entry.path)
+            exit_code += parse_dir(entry.path)
         elif entry.name.endswith('.ipynb'):
-            parse_notebook(entry.path)
+            exit_code += parse_notebook(entry.path, tag=directory.split('/')[-1], linkback=None)
+            
+    return exit_code
 
 
-def parse_notebook(path: str) -> None:
+def parse_notebook(path: str,
+                   tag: str,
+                   linkback: str) -> int:
     """
-        Review the specified notebook.
+        Review the specified notebook for conforming to the notebook template
+        and notebook authoring requirements.
         
-        path: The path to the notebook.
+            path: The path to the notebook.
+            tag: The associated tag
+            linkback: A link back to the web docs
+            
+        Returns the number of errors
     """
-    with open(path, 'r') as f:
-        try:
-            content = json.load(f)
-        except:
-            print("Corrupted notebook:", path)
-            return
-        
-        cells = content['cells']
+    notebook = Notebook(path)
+    
+    copyright = CopyrightRule()
+    notices = NoticesRule()
+    title = TitleRule()
+    links = LinksRule()
+    overview = OverviewRule()
+    objective = ObjectiveRule()
+    recommendations = RecommendationsRule()
+    dataset = DatasetRule()
+    costs = CostsRule()
+    setuplocal = SetupLocalRule()
+    helpers = HelpersRule()
+    installation = InstallationRule()
+    restart = RestartRule()
+    versions = VersionsRule()
+    beforebegin = BeforeBeginRule()
+    enableapis = EnableAPIsRule()
+    setupproject = SetupProjectRule()
+
         
-        NotebookRule.init()
+    copyright.validate(notebook)
+    notices.validate(notebook)
+    title.validate(notebook)
+    links.validate(notebook)
+    overview.validate(notebook)
+    objective.validate(notebook)
+    recommendations.validate(notebook)
+    dataset.validate(notebook)
+    costs.validate(notebook)
+    setuplocal.validate(notebook)
+    helpers.validate(notebook)
+    installation.validate(notebook)
+    restart.validate(notebook)
+    versions.validate(notebook)
+    beforebegin.validate(notebook)
+    enableapis.validate(notebook)
+    setupproject.validate(notebook)
+
+    
+    if objective.desc != '':
```

Comment:
This seems unrelated to validation and seems to be a side-effect for index creation.

Reply:
Yes, it is for indexing

Reply:
added comments to code